### PR TITLE
State

### PR DESCRIPTION
**Pull request by gabriel128**

### CODE REVIEW

File: src/Course/State.hs [hunk 1]
```diff
@@ -191,8 +170,33 @@ distinct =
...
+  | otherwise = computeHapiness x squareSumState
+      where
+        squareSumState = xs >>= (\a -> State (\s -> (a + 1, nextState s)))
+        nextState = digits . toInteger . sum . map (join (*))
+
+digits :: Integer -> List Int
+digits = reverse . map P.fromIntegral . unfoldr p
+  where
+    p 0 = Empty
+    p x = Full (x `mod` 10, x `div` 10)
```
Comment (JulianLeviston) [inline src/Course/State.hs:189]: nice :)

File: src/Course/State.hs [hunk 2]
```diff
@@ -191,8 +170,33 @@ distinct =
...
+
+digitsToNumber :: List Int -> Integer
+digitsToNumber = snd . foldRight reducer (1, 0)
```
Comment (JulianLeviston) [inline src/Course/State.hs]: hehe :) that's interesting :) ironically, we could use the state monad here, too, right? we want to keep and update a multiplier as we traverse the list, building up a result. :)

Comment (gabriel128) [reply src/Course/State.hs]: yep, that's true, let me refactor that just for fun

Comment (gabriel128) [reply src/Course/State.hs]:
```
digitsToNumber :: List Int -> Integer
digitsToNumber xs = exec (foldRight reducer (pure 1) xs) 0
  where
    reducer x oldState = oldState >>= (\mult -> State (\acc -> (mult * 10, toInteger x * mult + acc))  )

``` 
it was fun indeed

**Pull request merged**